Chap_API_Struct.tex

```diff
@@ -2151,9 +2151,118 @@ \subsubsection{Create the\refstruct{pmix_info_t} array of query qualifiers}
 
 
 %%%%%%%%%%%
-\section{Packing/Unpacking Types \& Structures}
+\subsection{Attribute registration structure}
+\declarestruct{pmix_regattr_t}
+
+The \refstruct{pmix_regattr_t} structure is used to register attribute support for a \ac{PMIx} function.
+
+\versionMarker{4.0}
+\cspecificstart
+\begin{codepar}
+typedef struct pmix_regattr \{
+    char *key;
+    pmix_data_type_t type;
+    char **valid_values;
+\} pmix_regattr_t;;
+\end{codepar}
+\cspecificend
+
+Characters in the key must be standard alphanumeric values supported by common utilities such as \textit{strcmp}. The \refarg{type} must be a \ac{PMIx} data type identifying the type of data associated with the \refarg{key} attribute. The \refarg{valid_values} field consists of a \code{NULL}-terminated array of strings identifying the valid values (or range of values) plus any other desired explanation.
```

Comment:
The `plus any other desired explanation` part is somewhat confusing for me. The use case I was thinking of is having a list of valid values that I can then turn around and use. But if that valid value has concatenated with it a description then I cannot do that without knowing the boundary.

I'm thinking that adding an explanation in that field is not preferred, and instead of having a separate field for documentation or leaving it off and requiring the RM to document any restrictions. Programmatically a description like that isn't useful except for printing back to the user.

Reply:
I think there are two use-cases here:

1. request from a tool (e.g., pattr) that displays the result to a human user. In this case, we want to be able to provide any additional description

1. request from a program that intends to parse the result. In this case, we want to clearly separate out and describe the accepted values in a machine-understandable way.

What if we add an attribute that requests the output be in machine-parsable form? This would require something other than just strings for valid values - e.g., the library currently returns "RANGE OF POSITIVE INTEGERS" for at least one attribute valid value, but would we want to declare that response as part of the standard? Or is it better to have attributes for minimum and maximum values that can include the data type (i.e., float vs int vs uint16_t)?

In other words, if we want this to be machine-parsable, then we have to be more descriptive of the valid values. Do we want to do that, or just support human-readable output?

Reply:
The more I think about it the more this sounds like the discussion around MCA parameters in OMPI. Originally available as strings to display via `ompi_info`, but later some were exposed more programatically for tools to play with.

Maybe for this first pass we focus on a `ompi_info` like tool to report out to the user, then think more about the latter case as an extension of this functionality.

With that in mind, I think having the `valid_values` be an array of descriptive strings is probably fine. If there was a way to express the separation between the value and the description, when necessary, I think that would be helpful. For example, something like:
```
valid_values[0] = strdup("values < -1 are reserved for future use");
valid_values[1] = strdup("\"-1\" indicates enable X with default values");
valid_values[2] = strdup("\"0\" indicates disable X");
valid_values[3] = strdup("values > 0 indicates enable X with that specific value");
```
So the `-1` and `0` cases are identified as specific values by using quotes (or some other unique character), but the rest express a general comment for the user about ranges of values.


Having an interface that can be inspected more programatically would be helpful if I were writing a program that wants to adapt to the availability of different attributes/values. But I think I'd want someone intending to do that to give some guidance on how to form up the interface. Maybe post this as an issue for a separate discussion.